server/src/db.rs

```diff
@@ -193,73 +302,140 @@ impl Db {
             .collect()
     }
 
-    /// Drops the specified chunk from the mutable buffer, returning
-    /// the dropped chunk.
-    pub async fn drop_mutable_buffer_chunk(
-        &self,
-        partition_key: &str,
-        chunk_id: u32,
-    ) -> Result<Arc<DBChunk>> {
-        self.mutable_buffer
+    /// Drops the specified chunk from the catalog and all storage systems
+    pub fn drop_chunk(&self, partition_key: &str, chunk_id: u32) -> Result<()> {
+        debug!(%partition_key, %chunk_id, "dropping chunk");
+
+        let mutable_buffer = self
+            .mutable_buffer
             .as_ref()
-            .context(DatatbaseNotWriteable)?
-            .drop_chunk(partition_key, chunk_id)
-            .map(|c| DBChunk::new_mb(c, partition_key, false))
-            .context(MutableBufferDrop)
-    }
+            .context(DatatbaseNotWriteable)?;
 
-    /// Drops the specified chunk from the read buffer, returning
-    /// the dropped chunk.
-    pub async fn drop_read_buffer_chunk(
-        &self,
-        partition_key: &str,
-        chunk_id: u32,
-    ) -> Result<Arc<DBChunk>> {
-        self.read_buffer
+        // Remove chunk from the catalog
+        let mut catalog = self.catalog.write();
+        let chunk_state = catalog
+            .chunk(partition_key, chunk_id)
+            .context(DroppingChunk {
+                partition_key,
+                chunk_id,
+            })?
+            .state();
+
+        // avoid dropping chunks that are actively being moved??  TODO
+        // it would be nicer to allow this to happen have the chunk
+        // migration logic cleanup afterwards so that users weren't
+        // prevented from dropping chunks due to background tasks
+        if chunk_state == ChunkState::Moving {
+            return DropMovingChunk {
+                partition_key,
+                chunk_id,
+                chunk_state,
+            }
+            .fail();
+        }
+
+        catalog
             .drop_chunk(partition_key, chunk_id)
-            .context(ReadBufferDrop)?;
+            .context(DroppingChunk {
+                partition_key,
+                chunk_id,
+            })?;
+
+        // clear it also from the read buffer / mutable buffer, if needed
+        let (drop_mb, drop_rb) = match chunk_state {
+            ChunkState::Open => (true, false),
+            ChunkState::Closing => (true, false),
+            ChunkState::Closed => (true, false),
+            ChunkState::Moving => (true, true), // data may be partially in read buffer
```

Comment:
Surely this should be the DropMovingChunk error above?

Reply:
yeah -- you are right that this should be caught by the check above. I'll update the comments

```suggestion
            ChunkState::Moving => (true, true), // not possible as `ChunkState::Moving` checked above
```